llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp

```diff
@@ -0,0 +1,1589 @@
+    return false;
+  }
+
+  return true;
+}
+
+bool RISCVVLOptimizer::tryReduceVL(MachineInstr &OrigMI) {
+  SetVector<MachineInstr *> Worklist;
+  Worklist.insert(&OrigMI);
+
+  bool MadeChange = false;
+  while (!Worklist.empty()) {
+    MachineInstr &MI = *Worklist.pop_back_val();
+    LLVM_DEBUG(dbgs() << "Try reduce VL for " << MI << "\n");
+    std::optional<Register> CommonVL;
+    bool CanReduceVL = true;
+    for (auto &UserOp : MRI->use_operands(MI.getOperand(0).getReg())) {
+      const MachineInstr &UserMI = *UserOp.getParent();
+      LLVM_DEBUG(dbgs() << "  Check user: " << UserMI << "\n");
+
+      // Instructions like reductions may use a vector register as a scalar
+      // register. In this case, we should treat it like a scalar register which
+      // does not impact the decision on whether to optimize VL.
+      if (isVectorOpUsedAsScalarOp(UserOp)) {
+        [[maybe_unused]] Register R = UserOp.getReg();
+        [[maybe_unused]] const TargetRegisterClass *RC = MRI->getRegClass(R);
+        assert(RISCV::VRRegClass.hasSubClassEq(RC) &&
+               "Expect LMUL 1 register class for vector as scalar operands!");
+        LLVM_DEBUG(dbgs() << "    Use this operand as a scalar operand\n");
+        continue;
+      }
+
+      if (!safeToPropgateVL(UserMI)) {
+        LLVM_DEBUG(dbgs() << "    Abort due to used by unsafe instruction\n");
+        CanReduceVL = false;
+        break;
+      }
+
+      // Tied operands might pass through.
+      if (UserOp.isTied()) {
+        LLVM_DEBUG(dbgs() << "    Abort due to user use it as tied operand\n");
+        CanReduceVL = false;
+        break;
+      }
+
+      const MCInstrDesc &Desc = UserMI.getDesc();
+      if (!RISCVII::hasVLOp(Desc.TSFlags) || !RISCVII::hasSEWOp(Desc.TSFlags)) {
+        LLVM_DEBUG(dbgs() << "    Abort due to lack of VL or SEW, assume that"
+                             " use VLMAX.\n");
+        CanReduceVL = false;
+        break;
+      }
+
+      unsigned VLOpNum = RISCVII::getVLOpNum(Desc);
+      const MachineOperand &VLOp = UserMI.getOperand(VLOpNum);
+      // Looking for a register VL that isn't X0.
+      if (!VLOp.isReg() || VLOp.getReg() == RISCV::X0) {
+        LLVM_DEBUG(dbgs() << "    Abort due to user use X0 as VL.\n");
+        CanReduceVL = false;
+        break;
+      }
+
+      if (!CommonVL) {
+        CommonVL = VLOp.getReg();
+      } else if (*CommonVL != VLOp.getReg()) {
+        LLVM_DEBUG(dbgs() << "    Abort due to users have different VL!\n");
+        CanReduceVL = false;
+        break;
+      }
+
+      // The SEW and LMUL of destination and source registers need to match.
+
+      // If the produced Dest is not a vector register, then it has no EEW or
+      // EMUL, so there is no need to check that producer and consumer LMUL and
+      // SEW match. We've already checked above that UserOp is a vector
+      // register.
+      if (!isVectorRegClass(MI.getOperand(0).getReg(), MRI)) {
```

Comment:
This can be hoisted out of the loop right?  Though actually, how does such a thing end up on the visit list at all?  This maybe implies a missing guard somewhere.

Reply:
At first glance, I agreed with the statement that we can hoist this out. But under closer inspection, I am not so sure.

> We've already checked above that UserOp is a vector register.

I think this part is key. If we hoist this check and continue in the outer loop, then we are not sure that UserOp is a vector register. I have to give some thought to whether we'd continue if both were non vector reg class.

Reply:
I gave some more thought to this. First I wondered whether it belonged in `isCandidate` since that is the guard to add to the worklist. I don't think it belongs as `isCandidate` since if an MI had VL and SEW operands but did not produce a vector instruction , we would want to reduce it.

For example:
```
a = ..., vl_op, sew_op
b = ..., vl_op, sew_op
scalar_register = some_vector_instr a, b, vl_op, sew_op
```
In this case, we would still want to try to reduce the VL here, despite `some_vector_instr` takes vector instructions and produces a scalar register.

However, I am not aware of any instruction that takes vectors and produces a scalar -- in fact this was explicitly avoided in the case of reductions. It could happen in the future, but I think we don't need to worry about it right now.

With this knowledge, this check is redundant since isCandidate checks for hasVLOp and hasSEWOp and we know that there are no instructions that have this property with a scalar destination. I will move this check to an assert in isCandidate.

Reply:
> However, I am not aware of any instruction that takes vectors and produces a scalar

vmv.x.s and vfmv.f.s

Reply:
Edit: I take that back, VFIRST is an example of such an instruction.

Reply:
Addressed in 583bb4919602068f9ba02e21503312b372f79ee8.